File: solutions/typescript/leap/1/leap.ts
/**
 * Determines whether the given year is a leap year in the Gregorian calendar.
 *
 * Rules:
 * - Every year divisible by 4 is a leap year,
 *   except years divisible by 100 are not leap years,
 *   unless they are also divisible by 400.
 *
 * @param year - The year to check
 * @returns True if leap year, false otherwise
 */
export function isLeap(year: number): boolean {
  // Years divisible by 400 are leap years
  if (year % 400 === 0) {
    return true;
  }
  // Years divisible by 100 (but not 400) are not leap years
  if (year % 100 === 0) {
    return false;
  }
  // Years divisible by 4 (but not 100) are leap years
  if (year % 4 === 0) {
    return true;
  }
  // All other years are not leap years
  return false;
}

File: solutions/python/leap/1/leap.py
def leap_year(year):
    # Check if the year is divisible by 4
    if year % 4 == 0:
        # Check if the year is divisible by 100
        if year % 100 == 0:
            # If divisible by 100, check if it's also divisible by 400
            if year % 400 == 0:
                return True  # It's a leap year
            else:
                return False  # It's not a leap year
        else:
            return True  # It's a leap year
    else:
        return False  # It's not a leap year